include/h2o/memory.h

```diff
@@ -301,6 +305,13 @@ void h2o_append_to_null_terminated_list(void ***list, void *element);
 
 /* inline defs */
 
+inline void *h2o_memcpy(void *dst, const void *src, size_t n)
+{
+    if (src)
```

Comment:
> I would like to see such case being reported. If we implement the check as if (src != NULL) memcpy(...), we will lose the chance of receiving a report for such case.

Ah, I see. I understand your concern. I was focused on this scenario:
```
void g(void)
{
    printf("called\n");
}

int f(void *dst, void *src, size_t n) {
    if (!src)
        g();
    memcpy(dst, src, n);
}
```

Since `src` is passed to `memcpy`, and that this would lead to UB, the compiler could in theory remove the call to `g()`. Now, to be fair, none of the compilers I've tried this with compiled the code away (icc, clang or gcc), but from what I understand, it could be possible.

Maybe we could have an assert that if `src == NULL`, then `n` must be `0`? I believe this would cover both cases.